Feature/dependancy injection

NathiGlucode opened the pull request:
- Implemented a central DIContainer using Swinject
- Registered CoreDataManager as a singleton
- Registered WeatherRepository and TaskRepository as singletons
- Registered LocationManager as a singleton service
- Registered WeatherViewModel and TaskViewModel with their dependencies
- Added convenience resolve methods

NathiGlucode committed:
Implemented swinject framework

- Implemented a central DIContainer using Swinject
- Registered CoreDataManager as a singleton
- Registered WeatherRepository and TaskRepository as singletons
- Registered LocationManager as a singleton service
- Registered WeatherViewModel and TaskViewModel with their dependencies
- Added convenience resolve methods

NathiGlucode committed:
Refactoring DIContainer

NathiGlucode committed:
Fixed Implementing swinject framework

- Implemented a central DIContainer using Swinject
- Registered CoreDataManager as a singleton
- Registered WeatherRepository and TaskRepository as singletons
- Registered LocationManager as a singleton service
- Registered WeatherViewModel and TaskViewModel with their dependencies
- Added convenience resolve methods

SthembisoGlucode commented on SwiftToDoList/Views/ToDoView.swift:
```diff
@@ -16,9 +16,9 @@ struct ToDoView: View {
-    init(repository: TaskRepositoryProtocol) {
-            _viewModel = StateObject(wrappedValue: TaskViewModel(repository: repository)) // this will create the viewmodel and pass the repository it received
-        }
+    init() {
```
You can remove the init completely but still pass in the viewModel where you register your ToDoView in your DIContainer, the view should not do this StateObject(wrappedValue: DIContainer.shared.resolveTaskViewModel()).

It should still update as expected

SthembisoGlucode commented on SwiftToDoList/Views/WeatherView.swift:
```diff
@@ -8,10 +8,11 @@
+    @StateObject private var viewModel : WeatherViewModel
+    
+    init() {
+        _viewModel = StateObject(wrappedValue: DIContainer.shared.resolveWeatherViewModel())
```
same here, let the DIContainer handle the registering and the resolving of all dependencies, we can pair If you wanna discuss this

NathiGlucode committed:
Implemented PR comments & WeatherViewModeling

- Removed init from WeatherView and TaskView
- Created WeatherViewModeling protocol
-Made WeatherView a generic struct that accepts a weather view model that conforms to WeatherViewModeling

NathiGlucode committed:
Implement PR comments & Created Widget extension

- DIcontainer handle registering
- Widget extension
- created widget core data manager
